contracts/broker/contracts/src/Broker.sol

```diff
@@ -0,0 +1,253 @@
+/*
+
+  Copyright 2019 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-exchange/contracts/src/interfaces/IExchange.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibAssetDataTransfer.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibFillResults.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
+import "@0x/contracts-utils/contracts/src/LibBytes.sol";
+import "@0x/contracts-utils/contracts/src/LibRichErrors.sol";
+import "@0x/contracts-utils/contracts/src/LibSafeMath.sol";
+import "@0x/contracts-utils/contracts/src/Refundable.sol";
+import "./interfaces/IBroker.sol";
+import "./interfaces/IPropertyValidator.sol";
+import "./libs/LibBrokerRichErrors.sol";
+
+
+// solhint-disable space-after-comma
+contract Broker is
+    Refundable,
+    IBroker
+{
+    bytes[] internal _cachedAssetData;
+    uint256 internal _cacheIndex;
+    address internal _sender;
+    address internal _EXCHANGE; // solhint-disable-line var-name-mixedcase
+
+    using LibSafeMath for uint256;
+    using LibBytes for bytes;
+    using LibAssetDataTransfer for bytes;
+
+    /// @param exchange Address of the 0x Exchange contract.
+    constructor (address exchange)
+        public
+    {
+        _EXCHANGE = exchange;
+    }
+
+    /// @dev A payable fallback function that makes this contract "payable". This is necessary to allow
+    ///      this contract to gracefully handle refunds from the Exchange.
+    function ()
+        external
+        payable
+    {} // solhint-disable-line no-empty-blocks
+
+    /// @dev The Broker implements the ERC1155 transfer function to be compatible with the ERC1155 asset proxy
+    /// @param from Since the Broker serves as the taker of the order, this should equal `address(this)`
+    /// @param to This should be the maker of the order.
+    /// @param amounts Should be an array of just one `uint256`, specifying the amount of the brokered assets to transfer.
+    /// @param data Encodes the validator contract address and any auxiliary data it needs for property validation.
+    function safeBatchTransferFrom(
+        address from,
+        address to,
+        uint256[] calldata /* ids */,
+        uint256[] calldata amounts,
+        bytes calldata data
+    )
+        external
+    {
+        if (from != address(this)) {
```

Comment:
I put this in to handle the case where e.g. both `makerFeeAssetData = takerAssetData`. Without this check, the Broker would send one of the taker's assets to the fee recipient, but this would be disguised as a maker fee

Reply:
Could this also be an issue for the taker fee? There's nothing explicitly preventing filling an order w/ a taker fee right?

Reply:
Yeah, I suppose so, but at least if the `takerFee` is non-zero the taker should expect to be sending _something_ to the maker, so it's kinda on them/the front-end to surface what that thing is. Also I can't really think of an easy way for the Broker to distinguish between a taker -> maker transfer vs a taker -> feeRecipient transfer